## ch01

<div style="text-align: right"> <b>Author : Kwang Myung Yu</b></div>
<div style="text-align: right\"> Initial upload: 2022.06.03 </div>
<div style="text-align: right\"> Last update: 2022.06.03</div>

In [1]:
import os
import sys
import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings; warnings.filterwarnings('ignore')
plt.style.use('ggplot')
%matplotlib inline

In [2]:
colors = ["#00798c", "#d1495b", '#edae49', '#66a182', '#4a4a4a',
          '#1a508b', '#e3120b', '#c5a880', '#9F5F80', '#6F9EAF',
          '#0278ae','#F39233', '#A7C5EB', '#54E346', '#ABCE74',
        '#d6b0b1', '#58391c', '#cdd0cb', '#ffb396', '#6930c3']
sns.color_palette(colors[:10])

[(0.0, 0.4745098039215686, 0.5490196078431373),
 (0.8196078431372549, 0.28627450980392155, 0.3568627450980392),
 (0.9294117647058824, 0.6823529411764706, 0.28627450980392155),
 (0.4, 0.6313725490196078, 0.5098039215686274),
 (0.2901960784313726, 0.2901960784313726, 0.2901960784313726),
 (0.10196078431372549, 0.3137254901960784, 0.5450980392156862),
 (0.8901960784313725, 0.07058823529411765, 0.043137254901960784),
 (0.7725490196078432, 0.6588235294117647, 0.5019607843137255),
 (0.6235294117647059, 0.37254901960784315, 0.5019607843137255),
 (0.43529411764705883, 0.6196078431372549, 0.6862745098039216)]

### 데이터 랭글링 : 자전거 대여 데이터셋

In [3]:
df = pd.read_csv('data/bike_rentals.csv')

In [4]:
df.head()

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
0        1  2011-01-01     1.0  0.0   1.0      0.0      6.0         0.0   
1        2  2011-01-02     1.0  0.0   1.0      0.0      0.0         0.0   
2        3  2011-01-03     1.0  0.0   1.0      0.0      1.0         1.0   
3        4  2011-01-04     1.0  0.0   1.0      0.0      2.0         1.0   
4        5  2011-01-05     1.0  0.0   1.0      0.0      3.0         1.0   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

#### 데이터 이해하기

In [5]:
df.shape

(731, 16)

In [6]:
df.describe()

instant      season          yr        mnth     holiday     weekday  \
count  731.000000  731.000000  730.000000  730.000000  731.000000  731.000000   
mean   366.000000    2.496580    0.500000    6.512329    0.028728    2.997264   
std    211.165812    1.110807    0.500343    3.448303    0.167155    2.004787   
min      1.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
25%    183.500000    2.000000    0.000000    4.000000    0.000000    1.000000   
50%    366.000000    3.000000    0.500000    7.000000    0.000000    3.000000   
75%    548.500000    3.000000    1.000000    9.750000    0.000000    5.000000   
max    731.000000    4.000000    1.000000   12.000000    1.000000    6.000000   

       workingday  weathersit        temp       atemp         hum   windspeed  \
count  731.000000  731.000000  730.000000  730.000000  728.000000  726.000000   
mean     0.682627    1.395349    0.495587    0.474512    0.627987    0.190476   
std      0.465773    0.544894    0.183094    0.163017    0.142331    0.077725   
min      0.000000    1.000000    0.059130    0.079070    0.000000    0.022392   
25%      0.000000    1.000000    0.336875    0.337794    0.521562    0.134494   
50%      1.000000    1.000000    0.499166    0.487364    0.627083    0.180971   
75%      1.000000    2.000000    0.655625    0.608916    0.730104    0.233218   
max      1.000000    3.000000    0.861667    0.840896    0.972500    0.507463   

            casual   registered          cnt  
count   731.000000   731.000000   731.000000  
mean    848.176471  3656.172367  4504.348837  
std     686.622488  1560.256377  1937.211452  
min       2.000000    20.000000    22.000000  
25%     315.500000  2497.000000  3152.000000  
50%     713.000000  3662.000000  4548.000000  
75%    1096.000000  4776.500000  5956.000000  
max    3410.000000  6946.000000  8714.000000

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    float64
 3   yr          730 non-null    float64
 4   mnth        730 non-null    float64
 5   holiday     731 non-null    float64
 6   weekday     731 non-null    float64
 7   workingday  731 non-null    float64
 8   weathersit  731 non-null    int64  
 9   temp        730 non-null    float64
 10  atemp       730 non-null    float64
 11  hum         728 non-null    float64
 12  windspeed   726 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(10), int64(5), object(1)
memory usage: 91.5+ KB


#### 결측치 처리

결측치 확인

In [8]:
df.isnull().sum()

instant       0
dteday        0
season        0
yr            1
mnth          1
holiday       0
weekday       0
workingday    0
weathersit    0
temp          1
atemp         1
hum           3
windspeed     5
casual        0
registered    0
cnt           0
dtype: int64

결측치 확인

In [9]:
df[df.isna().any(axis=1)]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
56        57  2011-02-26     1.0  0.0   2.0      0.0      6.0         0.0   
81        82  2011-03-23     2.0  0.0   3.0      0.0      3.0         1.0   
128      129  2011-05-09     2.0  0.0   5.0      0.0      1.0         1.0   
129      130  2011-05-10     2.0  0.0   5.0      0.0      2.0         1.0   
213      214  2011-08-02     3.0  0.0   8.0      0.0      2.0         1.0   
298      299  2011-10-26     4.0  0.0  10.0      0.0      3.0         1.0   
388      389  2012-01-24     1.0  1.0   1.0      0.0      2.0         1.0   
528      529  2012-06-12     2.0  1.0   6.0      0.0      2.0         1.0   
701      702  2012-12-02     4.0  1.0  12.0      0.0      0.0         0.0   
730      731  2012-12-31     1.0  NaN   NaN      0.0      1.0         0.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
56            1  0.282500  0.282192  0.537917        NaN     424        1545   
81            2  0.346957  0.337939  0.839565        NaN     203        1918   
128           1  0.532500  0.525246  0.588750        NaN     664        3698   
129           1  0.532500  0.522721       NaN   0.115671     694        4109   
213           1  0.783333  0.707071       NaN   0.205850     801        4044   
298           2  0.484167  0.472846  0.720417        NaN     404        3490   
388           1  0.342500  0.349108       NaN   0.123767     439        3900   
528           2  0.653333  0.597875  0.833333        NaN     477        4495   
701           2       NaN       NaN  0.823333   0.124379     892        3757   
730           2  0.215833  0.223487  0.577500   0.154846     439        2290   

      cnt  
56   1969  
81   2121  
128  4362  
129  4803  
213  4845  
298  3894  
388  4339  
528  4972  
701  4649  
730  2729

In [10]:
df[df.isna().any(axis=1)].shape

(10, 16)

In [11]:
df.isna().any(axis=1)

0      False
1      False
2      False
3      False
4      False
       ...  
726    False
727    False
728    False
729    False
730     True
Length: 731, dtype: bool

In [12]:
df[df.isnull().any(axis=1)].shape

(10, 16)

결측치 처리 1 : 중간값, 평균 값으로 대체

In [13]:
df['windspeed'] = df['windspeed'].fillna(df['windspeed'].median())

In [14]:
#확인
df.iloc[[56, 81]]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
56       57  2011-02-26     1.0  0.0   2.0      0.0      6.0         0.0   
81       82  2011-03-23     2.0  0.0   3.0      0.0      3.0         1.0   

    weathersit      temp     atemp       hum  windspeed  casual  registered  \
56           1  0.282500  0.282192  0.537917   0.180971     424        1545   
81           2  0.346957  0.337939  0.839565   0.180971     203        1918   

     cnt  
56  1969  
81  2121

결측치 처리2 : 중간 값이나 평균으로 그룹바이 하기
**중요**

In [15]:
df[df['hum'].isnull()]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
129      130  2011-05-10     2.0  0.0   5.0      0.0      2.0         1.0   
213      214  2011-08-02     3.0  0.0   8.0      0.0      2.0         1.0   
388      389  2012-01-24     1.0  1.0   1.0      0.0      2.0         1.0   

     weathersit      temp     atemp  hum  windspeed  casual  registered   cnt  
129           1  0.532500  0.522721  NaN   0.115671     694        4109  4803  
213           1  0.783333  0.707071  NaN   0.205850     801        4044  4845  
388           1  0.342500  0.349108  NaN   0.123767     439        3900  4339

In [16]:
# season으로 그룹바이한 중간 값을 얻는다.
df.groupby(['season']).median()

instant   yr  mnth  holiday  weekday  workingday  weathersit  \
season                                                                 
1.0       366.0  0.5   2.0      0.0      3.0         1.0         1.0   
2.0       308.5  0.5   5.0      0.0      3.0         1.0         1.0   
3.0       401.5  0.5   8.0      0.0      3.0         1.0         1.0   
4.0       493.0  0.5  11.0      0.0      3.0         1.0         1.0   

            temp     atemp       hum  windspeed  casual  registered     cnt  
season                                                                       
1.0     0.285833  0.282821  0.543750   0.202750   218.0      1867.0  2209.0  
2.0     0.562083  0.538212  0.646667   0.191546   867.0      3844.0  4941.5  
3.0     0.714583  0.656575  0.635833   0.165115  1050.5      4110.5  5353.5  
4.0     0.410000  0.409708  0.661042   0.167918   544.5      3815.0  4634.5

In [17]:
df.groupby(['season'])['hum'].transform('median') # 계절에 따른...

0      0.54375
1      0.54375
2      0.54375
3      0.54375
4      0.54375
        ...   
726    0.54375
727    0.54375
728    0.54375
729    0.54375
730    0.54375
Name: hum, Length: 731, dtype: float64

In [18]:
df['hum']= df['hum'].fillna(df.groupby('season')['hum'].transform('median'))

In [19]:
df.iloc[[129, 213, 388]]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
129      130  2011-05-10     2.0  0.0   5.0      0.0      2.0         1.0   
213      214  2011-08-02     3.0  0.0   8.0      0.0      2.0         1.0   
388      389  2012-01-24     1.0  1.0   1.0      0.0      2.0         1.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
129           1  0.532500  0.522721  0.646667   0.115671     694        4109   
213           1  0.783333  0.707071  0.635833   0.205850     801        4044   
388           1  0.342500  0.349108  0.543750   0.123767     439        3900   

      cnt  
129  4803  
213  4845  
388  4339

결측치 처리3 : 특정 행에서 평균 구하기

In [20]:
df[df['temp'].isna()]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
701      702  2012-12-02     4.0  1.0  12.0      0.0      0.0         0.0   

     weathersit  temp  atemp       hum  windspeed  casual  registered   cnt  
701           2   NaN    NaN  0.823333   0.124379     892        3757  4649

In [21]:
# 이전 시간대와 이후 시간대의 값으로 평균을 구해서 결측치를 채운다.

mean_temp = (df.iloc[700]['temp'] + df.iloc[702]['temp'])/2
mean_atemp = (df.iloc[700]['atemp'] + df.iloc[702]['atemp'])/2

df['temp'] = df['temp'].fillna(mean_temp)
df['atemp'] = df['atemp'].fillna(mean_atemp)

In [22]:
df.iloc[[701]]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
701      702  2012-12-02     4.0  1.0  12.0      0.0      0.0         0.0   

     weathersit      temp    atemp       hum  windspeed  casual  registered  \
701           2  0.375417  0.38635  0.823333   0.124379     892        3757   

      cnt  
701  4649

날짜 추정하기

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    float64
 3   yr          730 non-null    float64
 4   mnth        730 non-null    float64
 5   holiday     731 non-null    float64
 6   weekday     731 non-null    float64
 7   workingday  731 non-null    float64
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(10), int64(5), object(1)
memory usage: 91.5+ KB


In [24]:
df['dteday'] = pd.to_datetime(df['dteday'])

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   instant     731 non-null    int64         
 1   dteday      731 non-null    datetime64[ns]
 2   season      731 non-null    float64       
 3   yr          730 non-null    float64       
 4   mnth        730 non-null    float64       
 5   holiday     731 non-null    float64       
 6   weekday     731 non-null    float64       
 7   workingday  731 non-null    float64       
 8   weathersit  731 non-null    int64         
 9   temp        731 non-null    float64       
 10  atemp       731 non-null    float64       
 11  hum         731 non-null    float64       
 12  windspeed   731 non-null    float64       
 13  casual      731 non-null    int64         
 14  registered  731 non-null    int64         
 15  cnt         731 non-null    int64         
dtypes: datetime64[ns](1), floa

In [26]:
# 또다른 방법

df['dteday'].apply(pd.to_datetime, infer_datetime_format=True, errors='coerce')

0     2011-01-01
1     2011-01-02
2     2011-01-03
3     2011-01-04
4     2011-01-05
         ...    
726   2012-12-27
727   2012-12-28
728   2012-12-29
729   2012-12-30
730   2012-12-31
Name: dteday, Length: 731, dtype: datetime64[ns]

In [27]:
df[df['mnth'].isnull()]

instant     dteday  season  yr  mnth  holiday  weekday  workingday  \
730      731 2012-12-31     1.0 NaN   NaN      0.0      1.0         0.0   

     weathersit      temp     atemp     hum  windspeed  casual  registered  \
730           2  0.215833  0.223487  0.5775   0.154846     439        2290   

      cnt  
730  2729

In [28]:
import datetime as dt

In [29]:
df['mnth'] = df['dteday'].dt.month

In [30]:
# 연도는 이전것을 참고하여 입력하자.

In [31]:
df.iloc[[728, 729, 730]]

instant     dteday  season   yr  mnth  holiday  weekday  workingday  \
728      729 2012-12-29     1.0  1.0    12      0.0      6.0         0.0   
729      730 2012-12-30     1.0  1.0    12      0.0      0.0         0.0   
730      731 2012-12-31     1.0  NaN    12      0.0      1.0         0.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
728           2  0.253333  0.242400  0.752917   0.124383     159        1182   
729           1  0.255833  0.231700  0.483333   0.350754     364        1432   
730           2  0.215833  0.223487  0.577500   0.154846     439        2290   

      cnt  
728  1341  
729  1796  
730  2729

In [32]:
df.loc[730, 'yr'] = 1.0

In [33]:
df.iloc[[728, 729, 730]]

instant     dteday  season   yr  mnth  holiday  weekday  workingday  \
728      729 2012-12-29     1.0  1.0    12      0.0      6.0         0.0   
729      730 2012-12-30     1.0  1.0    12      0.0      0.0         0.0   
730      731 2012-12-31     1.0  1.0    12      0.0      1.0         0.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
728           2  0.253333  0.242400  0.752917   0.124383     159        1182   
729           1  0.255833  0.231700  0.483333   0.350754     364        1432   
730           2  0.215833  0.223487  0.577500   0.154846     439        2290   

      cnt  
728  1341  
729  1796  
730  2729

In [34]:
# 다 사용한 날짜 컬럼 삭제
df = df.drop('dteday', axis = 1)

In [35]:
df.head()

instant  season   yr  mnth  holiday  weekday  workingday  weathersit  \
0        1     1.0  0.0     1      0.0      6.0         0.0           2   
1        2     1.0  0.0     1      0.0      0.0         0.0           2   
2        3     1.0  0.0     1      0.0      1.0         1.0           1   
3        4     1.0  0.0     1      0.0      2.0         1.0           1   
4        5     1.0  0.0     1      0.0      3.0         1.0           1   

       temp     atemp       hum  windspeed  casual  registered   cnt  
0  0.344167  0.363625  0.805833   0.160446     331         654   985  
1  0.363478  0.353739  0.696087   0.248539     131         670   801  
2  0.196364  0.189405  0.437273   0.248309     120        1229  1349  
3  0.200000  0.212122  0.590435   0.160296     108        1454  1562  
4  0.226957  0.229270  0.436957   0.186900      82        1518  1600

### 회귀 모델 만들기

#### 자전거 대여 예측하기

불필요한 컬럼 일단 제거

In [36]:
df = df.drop(['casual', 'registered'], axis = 1)

데이터 일단 저장

In [37]:
df.to_csv('data/bike_rentals_cleaned.csv', index = False)

피처, 타겟 준비

In [38]:
X = df.iloc[:, :-1]
y = df.iloc[:, -1]

In [39]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2)

모델 라이브러리 import 

In [40]:
import xgboost as xgb
xgb.set_config(verbosity = 0)

from sklearn.linear_model import LinearRegression

#### 선형회귀 모델

In [41]:
lin_reg = LinearRegression()

lin_reg.fit(X_train, y_train)
y_pred = lin_reg.predict(X_test)

평가

In [42]:
from sklearn.metrics import mean_squared_error

In [43]:
mse = mean_squared_error(y_test, y_pred)

In [44]:
rmse = np.sqrt(mse)

In [45]:
print('RMSE: %0.2f'%(rmse))

RMSE: 898.21


결과 분석 : RMSE가 좋은 것인가?
    
RMSE로 성능을 평가하려면 타겟 데이터를 확인해봐야 한다.

In [46]:
df['cnt'].describe()

count     731.000000
mean     4504.348837
std      1937.211452
min        22.000000
25%      3152.000000
50%      4548.000000
75%      5956.000000
max      8714.000000
Name: cnt, dtype: float64

평균 4504 대비 약 20% 넘게 오차가 발생한 것이다.  
좋지도 나쁘지도 않는 결과이다.

#### XGBRegressor

In [47]:
from xgboost import XGBRegressor

In [48]:
xg_reg = XGBRegressor()

xg_reg.fit(X_train, y_train)
y_pred = xg_reg.predict(X_test)

In [49]:
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

In [50]:
print('RMSE: %0.2f'% (rmse))

RMSE: 705.11


### 교차 검증

In [51]:
from sklearn.model_selection import cross_val_score

선형회귀 교차검증

In [52]:
model = LinearRegression()

scores = cross_val_score(model, X, y, scoring='neg_mean_squared_error', cv = 10)

In [53]:
scores

array([ -254023.184597  ,  -706523.99840198, -1301612.08252473,
        -530554.28645016,  -409851.78535971,  -940801.36212315,
       -1284705.47755477, -1569626.37894129, -1176435.35983868,
       -2031553.62469152])

In [54]:
rmse = np.sqrt(-scores)

In [55]:
rmse

array([ 504.00712753,  840.54981911, 1140.88215102,  728.39157494,
        640.19667709,  969.9491544 , 1133.44848915, 1252.84730871,
       1084.63604948, 1425.32579598])

In [56]:
# 평균 제곱근 오차를 출력합니다.
print('회귀 rmse:', np.round(rmse, 2))

# 평균 점수를 출력합니다.
print('RMSE 평균: %0.2f' % (rmse.mean()))

회귀 rmse: [ 504.01  840.55 1140.88  728.39  640.2   969.95 1133.45 1252.85 1084.64
 1425.33]
RMSE 평균: 972.02


In [57]:
# 같은 결과

-np.mean(cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=10))

972.023414741928

crossvalidate : cross_val_score는 cross_validate 함수를 랩핑한 것이다. crossvalidate를 사용하면 여러개의 지표를 동시에 사용할 수도 있다.

In [58]:
from sklearn.model_selection import cross_validate

In [59]:
cv_results = cross_validate(model, X, y , scoring='neg_root_mean_squared_error', cv=10)

In [60]:
cv_results

{'fit_time': array([0.01097274, 0.00897431, 0.00897622, 0.00698066, 0.00498819,
        0.0059855 , 0.00498605, 0.00598383, 0.00797749, 0.00698042]),
 'score_time': array([0.00598216, 0.00598454, 0.00398898, 0.00399065, 0.00398684,
        0.00299215, 0.0039885 , 0.00299287, 0.00299263, 0.00299215]),
 'test_score': array([ -504.00712753,  -840.54981911, -1140.88215102,  -728.39157494,
         -640.19667709,  -969.9491544 , -1133.44848915, -1252.84730871,
        -1084.63604948, -1425.32579598])}

- 더 많은 결과를 보여준다.

In [61]:
-np.mean(cv_results['test_score'])

972.023414741928

XGBoost 교차검증

In [62]:
model = XGBRegressor(objective='reg:squarederror')

In [63]:
model

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, gamma=None,
             gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, n_estimators=100, n_jobs=None,
             num_parallel_tree=None, predictor=None, random_state=None,
             reg_alpha=None, reg_lambda=None, ...)

In [64]:
scores = cross_val_score(model, X, y, scoring='neg_mean_squared_error', cv = 10)

In [65]:
rmse = np.sqrt(-scores)

In [66]:
rmse

array([ 717.64596522,  692.80382091,  520.70004518,  737.67570245,
        835.96102726, 1006.23650513,  991.34204876,  747.61229618,
        891.99416714, 1731.12815106])

In [67]:
# 평균 제곱근 오차를 출력합니다.
print('회귀 rmse:', np.round(rmse, 2))

# 평균 점수를 출력합니다.
print('RMSE 평균: %0.2f' % (rmse.mean()))

회귀 rmse: [ 717.65  692.8   520.7   737.68  835.96 1006.24  991.34  747.61  891.99
 1731.13]
RMSE 평균: 887.31


### 분류 모델 만들기

#### 데이터 랭글링

In [68]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data')

In [69]:
df.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

헤더가 없는 데이터셋을 로드하자.

In [70]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', header=None)

df.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [72]:
df.columns =  ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation',
                  'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 
                   'income']

df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

결측치 확인

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [74]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

카테고리 컬럼 처리

In [75]:
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [76]:
# education 컬럼 삭제
df = df.drop('education', axis = 1)

In [77]:
df.head()

age          workclass  fnlwgt  education-num       marital-status  \
0   39          State-gov   77516             13        Never-married   
1   50   Self-emp-not-inc   83311             13   Married-civ-spouse   
2   38            Private  215646              9             Divorced   
3   53            Private  234721              7   Married-civ-spouse   
4   28            Private  338409             13   Married-civ-spouse   

           occupation    relationship    race      sex  capital-gain  \
0        Adm-clerical   Not-in-family   White     Male          2174   
1     Exec-managerial         Husband   White     Male             0   
2   Handlers-cleaners   Not-in-family   White     Male             0   
3   Handlers-cleaners         Husband   Black     Male             0   
4      Prof-specialty            Wife   Black   Female             0   

   capital-loss  hours-per-week  native-country  income  
0             0              40   United-States   <=50K  
1             0              13   United-States   <=50K  
2             0              40   United-States   <=50K  
3             0              40   United-States   <=50K  
4             0              40            Cuba   <=50K

In [85]:
(df.dtypes =='object')

age               False
workclass          True
fnlwgt            False
education-num     False
marital-status     True
occupation         True
relationship       True
race               True
sex                True
capital-gain      False
capital-loss      False
hours-per-week    False
native-country     True
income             True
dtype: bool

In [88]:
df.select_dtypes(include=['object']).head()

workclass       marital-status          occupation    relationship  \
0          State-gov        Never-married        Adm-clerical   Not-in-family   
1   Self-emp-not-inc   Married-civ-spouse     Exec-managerial         Husband   
2            Private             Divorced   Handlers-cleaners   Not-in-family   
3            Private   Married-civ-spouse   Handlers-cleaners         Husband   
4            Private   Married-civ-spouse      Prof-specialty            Wife   

     race      sex  native-country  income  
0   White     Male   United-States   <=50K  
1   White     Male   United-States   <=50K  
2   White     Male   United-States   <=50K  
3   Black     Male   United-States   <=50K  
4   Black   Female            Cuba   <=50K

인코딩

In [89]:
df = pd.get_dummies(df)

In [90]:
df.head()

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0             0                       0                     0   
1             0                       0                     0   
2             0                       0                     0   
3             0                       0                     0   
4             0                       0                     0   

   workclass_ Never-worked  ...  native-country_ Scotland  \
0                        0  ...                         0   
1                        0  ...                         0   
2                        0  ...                         0   
3                        0  ...                         0   
4                        0  ...                         0   

   native-country_ South  native-country_ Taiwan  native-country_ Thailand  \
0                      0                       0                         0   
1                      0                       0                         0   
2                      0                       0                         0   
3                      0                       0                         0   
4                      0                       0                         0   

   native-country_ Trinadad&Tobago  native-country_ United-States  \
0                                0                              1   
1                                0                              1   
2                                0                              1   
3                                0                              1   
4                                0                              0   

   native-country_ Vietnam  native-country_ Yugoslavia  income_ <=50K  \
0                        0                           0              1   
1                        0                           0              1   
2                        0                           0              1   
3                        0                           0              1   
4                        0                           0              1   

   income_ >50K  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 94 columns]

In [91]:
df.head()

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0             0                       0                     0   
1             0                       0                     0   
2             0                       0                     0   
3             0                       0                     0   
4             0                       0                     0   

   workclass_ Never-worked  ...  native-country_ Scotland  \
0                        0  ...                         0   
1                        0  ...                         0   
2                        0  ...                         0   
3                        0  ...                         0   
4                        0  ...                         0   

   native-country_ South  native-country_ Taiwan  native-country_ Thailand  \
0                      0                       0                         0   
1                      0                       0                         0   
2                      0                       0                         0   
3                      0                       0                         0   
4                      0                       0                         0   

   native-country_ Trinadad&Tobago  native-country_ United-States  \
0                                0                              1   
1                                0                              1   
2                                0                              1   
3                                0                              1   
4                                0                              0   

   native-country_ Vietnam  native-country_ Yugoslavia  income_ <=50K  \
0                        0                           0              1   
1                        0                           0              1   
2                        0                           0              1   
3                        0                           0              1   
4                        0                           0              1   

   income_ >50K  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 94 columns]

피처와 타깃 분리

In [92]:
#'income_ <=50K' 열을 삭제합니다.
df = df.drop('income_ <=50K', axis=1)

In [93]:
X = df.iloc[:, :-1]
y = df.iloc[:, -1]

#### 로지스틱 회귀

In [94]:
def cross_val(classifier, num_splits = 10):
    model = classifier
    scores = cross_val_score(model, X, y, cv=num_splits)
    
    # 정확도 출력
    print('정확도:', np.round(scores, 2))

    # 평균 정확도 출력
    print('평균 정확도: %0.2f' % (scores.mean()))

In [96]:
from sklearn.linear_model import LogisticRegression

In [97]:
# LogisticRegression으로 cross_val 함수를 호출합니다.
cross_val(LogisticRegression())

정확도: [0.8  0.8  0.79 0.8  0.79 0.81 0.79 0.79 0.8  0.8 ]
평균 정확도: 0.80


#### XGBClassifier

In [98]:
from xgboost import XGBClassifier

In [99]:
cross_val(XGBClassifier(n_estimators = 5))

정확도: [0.85 0.86 0.87 0.85 0.86 0.86 0.86 0.87 0.86 0.86]
평균 정확도: 0.86
